driver/fvGFS/fv_ufs_restart_io.F90

```diff
@@ -0,0 +1,544 @@
+           core_center_var3_names(n) = 'W'; n=n+1
+           core_center_var3_names(n) = 'DZ'; n=n+1
+           if ( atm%flagstruct%hybrid_z ) then
+               core_center_var3_names(n) = 'ze0'; n=n+1
+           endif
+      else !the restart file has the non-hydrostatic variables
+           core_center_var3_names(n) = 'W'; n=n+1
+           core_center_var3_names(n) = 'DZ'; n=n+1
+           if ( atm%flagstruct%hybrid_z ) then
+               core_center_var3_names(n) = 'ze0'; n=n+1
+           endif
+      endif
+   endif
+   core_center_var3_names(n) = 'T'; n=n+1
+   core_center_var3_names(n) = 'delp'; n=n+1
+   !--- include agrid winds in restarts for use in data assimilation
+   if (atm%flagstruct%agrid_vel_rst) then
+     core_center_var3_names(n) = 'ua'; n=n+1
+     core_center_var3_names(n) = 'va'; n=n+1
+   endif
+
+   core_center_var2_names(1) = 'phis'
+
+   ! srf_wnd
+   nvar2d_srf_wnd = 2
+   allocate (srf_wnd_var2(nx,ny,nvar2d_srf_wnd), srf_wnd_var2_names(nvar2d_srf_wnd))
+   srf_wnd_var2_names(1) = 'u_srf'
+   srf_wnd_var2_names(2) = 'v_srf'
+
+   ! tracers
+   ntprog = size(Atm%q,4)
+   ntdiag = size(Atm%qdiag,4)
+   nvar3d_tracers = ntprog+ntdiag
+   tracers_zsize = size(Atm%q,3)
+   allocate (tracers_var3(nx,ny,tracers_zsize,nvar3d_tracers), tracers_var3_names(nvar3d_tracers))
+
+   do nt = 1, ntprog
+      call get_tracer_names(MODEL_ATMOS, nt, tracer_name)
+      tracers_var3_names(nt) = tracer_name
+   enddo
+   do nt = ntprog+1, nvar3d_tracers
+      call get_tracer_names(MODEL_ATMOS, nt, tracer_name)
+      tracers_var3_names(nt) = tracer_name
+   enddo
+
+ end subroutine fv_dyn_restart_register
+
+ subroutine fv_dyn_restart_output(Atm, timestamp)
+
+   implicit none
+
+   type(fv_atmos_type), intent(in) :: Atm
+   character(len=*), intent(in) :: timestamp
+
+   integer :: isc, iec, jsc, jec, nx, ny
+   integer :: n, nt
+   integer :: mlon, mlat, nx_c, ny_c
+
+   isc = Atm%bd%isc
+   iec = Atm%bd%iec
+   jsc = Atm%bd%jsc
+   jec = Atm%bd%jec
+   nx  = (iec - isc + 1)
+   ny  = (jec - jsc + 1)
+
+   call atmosphere_resolution (mlon, mlat, global=.true.)
+   nx_c = nx
+   ny_c = ny
+   if (iec == mlon) then
+      ! we are on task at the 'east' edge of the cubed sphere face or regional domain
+      ! corner arrays should have one extra element in 'i' direction
+      nx_c = nx + 1
+   end if
+   if (jec == mlat) then
+      ! we are on task at the 'north' edge of the cubed sphere face or regional domain
+      ! corner arrays should have one extra element in 'j' direction
+      ny_c = ny + 1
+   end if
+
+   ! ---- fv_core.res
+
+   core_north_var3(:,:,:,1) = Atm%u(isc:iec,jsc:(jsc+ny_c-1),:)
+   core_east_var3 (:,:,:,1) = Atm%v(isc:(isc+nx_c-1),jsc:jec,:)
+
+   n = 1
+   if (.not.atm%flagstruct%hydrostatic) then
+      if (atm%flagstruct%make_nh) then ! hydrostatic restarts dont have these variables
+           ! core_center_var3(n) = 'w'; n=n+1
```

Comment:
Not sure if I understand this if statement (and comments) . Isn't line 208 said that this is non-hydrostatic? Can we have:
```
if (.not. atm%flagstruct%make_nh) then
 lines 216-220
endif
```

Reply:
I copied the logic from:

https://github.com/NOAA-GFDL/GFDL_atmos_cubed_sphere/blob/dev/emc/tools/fv_io.F90#L301-L315